tardis/visualization/tools/liv_plot.py

```diff
@@ -0,0 +1,532 @@
+                    r_last_interaction = (
+                        g_df["last_interaction_in_r"].values * u.cm
+                    )
+                    v_last_interaction = (
+                        r_last_interaction / self.time_explosion
+                    ).to("km/s")
+                    full_v_last.extend(v_last_interaction)
+            if full_v_last:
+                plot_data.append(full_v_last)
+                plot_colors.append(self._color_list[species_counter])
+                species_counter += 1
+
+        return plot_data, plot_colors
+
+    def _prepare_plot_data(
+        self, packets_mode, species_list, cmapname, num_bins, nelements
+    ):
+        """
+        Prepare data and settings required for generating a plot.
+
+        This method handles the common logic for preparing data and settings
+        needed to generate both matplotlib and plotly plots. It parses the species
+        list, generates color labels and colormap, and bins the velocity data.
+
+        Parameters
+        ----------
+        packets_mode : str
+            Packet mode, either 'virtual' or 'real'.
+        species_list : list of str
+            List of species to plot. Species can be specified as an ion
+            (e.g., Si II), an element (e.g., Si), a range of ions (e.g., Si I-V),
+            or any combination of these.
+        cmapname : str
+            Name of the colormap to use. A specific colormap can be chosen, such
+            as "jet", "viridis", "plasma", etc.
+        num_bins : int, optional
+            Number of bins for regrouping within the same range. If None,
+            no regrouping is done.
+
+        Raises
+        ------
+        ValueError
+            If no species are provided for plotting, or if no valid species are
+            found in the model.
+
+        Returns
+        -------
+        plot_data : list
+            List of velocity data for each species.
+        plot_colors : list
+            List of colors corresponding to each species.
+        new_bin_edges : np.ndarray
+            Array of bin edges for the velocity data.
+        """
+        if species_list is None:
+            # Extract all unique elements from the packets data
+            species_in_model = np.unique(
+                self.data[packets_mode]
+                .packets_df_line_interaction["last_line_interaction_species"]
+                .values
+            )
+            species_list = [
+                f"{atomic_number2element_symbol(specie // 100)}"
+                for specie in species_in_model
+            ]
+        self._parse_species_list(species_list, packets_mode, nelements)
+        species_in_model = np.unique(
+            self.data[packets_mode]
+            .packets_df_line_interaction["last_line_interaction_species"]
+            .values
+        )
+        if self._species_list is None or not self._species_list:
+            raise ValueError("No species provided for plotting.")
```

Comment:
I believe this is no longer the case since you updated the logic to handle `None`. But you know the edge cases better than me, so only remove it if you are sure